lms/validation/_bearer_token.py

```diff
@@ -0,0 +1,219 @@
+    request's authorization param using webargs::
+
+        >>> from webargs.pyramidparser import parser
+        >>> parser.parse(s, request)
+        LTIUser(user_id='...', oauth_consumer_key='...', ...)
+    """
+
+    user_id = marshmallow.fields.Str(required=True)
+    oauth_consumer_key = marshmallow.fields.Str(required=True)
+
+    class Meta:
+        """Marshmallow options for this schema."""
+
+        # Silence a strict=False deprecation warning from marshmallow.
+        # TODO: Remove this once we've upgraded to marshmallow 3.
+        strict = True
+
+    def __init__(self, request):
+        super().__init__()
+        self._request = request
+        # Storing context needed for serialization or deserialization in
+        # self.context is a marshmallow convention.
+        self.context = {"secret": request.registry.settings["jwt_secret"]}
+
+    def authorization_param(self, lti_user):
+        """
+        Return ``lti_user`` serialized into an authorization param.
+
+        Returns a ``"Bearer: <ENCODED_JWT>"`` string suitable for use as the
+        value of an authorization param.
+
+        :arg lti_user: the LTI user to return an auth param for
+        :type lti_user: LTIUser
+
+        :rtype: str
+        """
+        return self.dump(lti_user).data["authorization"]
+
+    def lti_user(self):
+        """
+        Return an :class:`~lms.values.LTIUser` from the request's authorization param.
+
+        Verifies the signature and timestamp of the JWT in the request's
+        authorization param, decodes the JWT, validates the JWT's payload, and
+        returns an :class:`~lms.values.LTIUser` from the payload.
+
+        The authorization param must be in an HTTP header
+        ``Authorization: Bearer <ENCODED_JWT>`` in the request. In the future
+        we may add support for reading the authorization param from other parts
+        of the request, such as from form fields or JSON parameters.
+
+        :raise ExpiredSessionTokenError: if the request's Authorization header
+          contains an expired JWT
+        :raise MissingSessionTokenError: if the request does not contain an
+          Authorization header
+        :raise InvalidSessionTokenError: if the request contains an invalid
+          Authorization header. For example if the Authorization header
+          contains an invalid JWT, or if it doesn't have the required
+          ``"Bearer <ENCODED_JWT>"`` format.
+        :raise ValidationError: if the JWT's payload is invalid, for example if
+          it's missing a required parameter
+
+        :rtype: LTIUser
+        """
+        try:
+            return parser.parse(self, self._request, locations=["headers"])
+        except HTTPUnprocessableEntity as error:
+            try:
+                authorization_error_message = error.json["authorization"][0]
+            except KeyError:
+                exc_class = ValidationError
+            else:
```

Comment:
It's so that the `else` is _only_ executed if the `try` above completes without raising. If the `try` raises then the `except` will be executed but the `else` will not be. The `else` could go at the bottom of the `try` block but it's best to minimize how much code is inside a `try`